openflagr/flagr
# Customizable validation rules for enabled flags

## Description
This changes adds customizable validation rules for _enabled_ flags using https://github.com/Knetic/govaluate.  The rules are specified via new environment variables in the `env.go`; a list of govaluate boolean expressions to be evaluated for flags (a `[]string`), and a connecting operation (either `OR` or `AND`).  When someone uses the API operations `setFlagEnabled` or `putFlag` and the flag is `enabled=true`, we convert the flag into a `map[string]interface{}` so that the properties of the flag are parameters for the govaluate expression.  Then, we iterate through the govaluate expressions in order; if the boolean result of the govaluate expression is `true` then the flag has passed that rule.  Depending on the operation, at least one or all the rules must pass.  If the flag doesn't meet this criteria, a 400 request is returned to the client and the flag update is not saved. 

## Motivation and Context

Our team is trying to keep more aware of how flags are being used on our project.  Specifically we want to associated a flag with an owning project in our JIRA system, so we can provide the teams some info about how the flag is being used (e.g. if the flag in practice always returns the same segment for a period of time).  We landed on using a tag to track this relationship, and we wanted to automate user feedback making sure they assign a tag when the flag is active.  Doing this validation only when a flag is set to enabled seemed like the best way; the user can create and configure the flag without validation up until they enable it for actual use. 


I mentioned this change in this issue at the end: https://github.com/checkr/flagr/issues/428

## How Has This Been Tested?




I initially prototyped using govaluate for this usecase here: https://github.com/sesquipedalian-dev/govaluate-experiment/blob/main/main.go

In addition to adding unit tests covering the new validation rules, I followed a manual test procedure: 
0) environment: Macbook Pro, OSX 10.15.5, chrome browser, go 1.15.6
1) `make run_ui`
2) `FLAGR_ENABLED_FLAG_VALIDATION_RULES='any("regexMatch(Value, \"^JIRA:[a-zA-Z]{3}[a-zA-Z]*$\")", Tags)' make run`
3) in a browser at `localhost:8080` create a new flag with description 'New flag'
4) in the browser open developer tools to follow along with the network calls
5) Click the toggle button to enable the flag, observe you get an error message and the network tab indicates a 400
![Screen Shot 2021-01-25 at 10 47 34 AM](https://user-images.githubusercontent.com/23544853/105748967-08f75e00-5f00-11eb-960e-9300d43d4bac.png)
6) Reload the page (verify the flag in the DB is not actually enabled) - small UI bug here in that the toggle remains 'enabled' in the UI even though the network call failed
7) Add a tag to the flag matching the critera (e.g. `JIRA:EPLT`)
8) Click the toggle button to enable, observe no error message this time and the network call has a 200 status
9) Remove the tag from the flag
10) change the flag description by adding some characters, and click 'save'.  Observe you again get an error message as in step (5)
11)Add the tag back in and save again, observe no error message

I didn't test other areas of the code too much manually.

## Types of changes


## Checklist:
